src/Workspaces/Core/Portable/Execution/AssetStorages.Storage.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Serialization;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Execution
+{
+    /// <summary>
+    /// This is collection of root checksum tree node
+    /// </summary>
+    internal partial class AssetStorages
+    {
+        /// <summary>
+        /// root tree node of checksum tree
+        /// </summary>
+        public sealed class Storage
+        {
+            // cache to remove lambda allocation
+            private readonly static Func<ConcurrentDictionary<Checksum, CustomAsset>> s_additionalAssetsCreator = () => new ConcurrentDictionary<Checksum, CustomAsset>(concurrencyLevel: 2, capacity: 10);
+
+            private readonly Serializer _serializer;
+
+            // additional assets that is not part of solution but added explicitly
+            private ConcurrentDictionary<Checksum, CustomAsset> _additionalAssets;
```

Comment:
Essentially because we get continually burned by it :)

And even if this code is ok, i'm worried about people who maintain it and then accidently trip over one of it's landmines.  reader/writer locks are simple, fast and solve the problem great when you have many readers and few writers.  ConcurrentDictionary is really when you have a huge amount of writers as well, and you really can't even take the overhead of serializing out any work.